### PR TITLE
LibELF+LibC: Add support for IFUNCs / Implement an optimized memset routine for x86-64

### PR DESCRIPTION
**Pull request by BertalanD**
The fun parts are in the following two commits:

**LibELF: Add support for IFUNCs**
IFUNC is a GNU extension to the ELF standard that allows a function to
have multiple implementations. A resolver function has to be called at
load time to choose the right one to use. The PLT will contain the entry
to the resolved function, so branching and more indirect jumps can be
avoided at run-time.

This mechanism is usually used when a routine can be made faster using
CPU features that are available in only some models, and a fallback
implementation has to exist for others.

We will use this feature to have two separate memset implementations for
CPUs with and without ERMS (Enhanced REP MOVSB/STOSB) support.

**LibC: Implement a faster memset routine for x86-64 in assembly**
This commit addresses the following shortcomings of our current, simple
and elegant memset function:
- REP STOSB/STOSQ has considerable startup overhead, it's impractical to
  use for smaller sizes.
- Up until very recently, AMD CPUs didn't have support for "Enhanced REP
  MOVSB/STOSB", so it performed pretty poorly on them.

With this commit applied, I could measure a ~5% decrease in `test-js`'s
runtime when I used qemu's TCG backend. The implementation is based on
the following article from Microsoft:

https://msrc-blog.microsoft.com/2021/01/11/building-faster-amd64-memset-routines

Two versions of the routine are implemented: one that uses the ERMS
extension mentioned above, and one that performs plain SSE stores. The
version appropriate for the CPU is selected at load time using an IFUNC.

**Note that IFUNC support was added very recently to the toolchain, so a
rebuild will be required to get rid of the "'ifunc' is not supported on
this target" message.**

`IFUNC`s change the ELF's ABI to the GNU OS ABI, so the commit from #12881 that relaxes the kernel's ABI restriction is included in this PR.

### CODE REVIEW

**Comment by BertalanD** [comment]
This PR is stuck on a pretty nasty issue.

LibC's dynamic relocations have to be performed before IFUNC resolution because the resolver depends on the addresses of the `memset_sse2` and `memset_sse2_erms` functions pointing to the right place.

This does not happen with the Clang toolchain because libc wants to load `libunwind` dynamically, so LibC is not the first one in the topological order. Using static `linbunwind` is not an option because it wants symbols from LibM and LibPthread.

My idea is to unify all the C runtime libraries into one like musl does.

**Comment by BertalanD** [comment]
For the time being, let's just resolve `memset` at runtime, the LibC changes needed for statically linking libunwind can come later.

I'm opening up this PR for review.

File: Userland/Libraries/LibC/arch/x86_64/memset.S [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    rep  stosb
+    mov  rax, rsi
+
+    ret
+
+.global  memset_sse2
+.type    memset_sse2, @function
+.p2align 4
+
+memset_sse2:
```
Comment (ADKaster) [inline Userland/Libraries/LibC/arch/x86_64/memset.S:75]: I don't know enough x86 asm to know the answer for sure, but does this implementation properly push/preserve all callee-saved registers? These routines have a suspicious lack of function prologue and epilogue.

Comment (BertalanD) [reply Userland/Libraries/LibC/arch/x86_64/memset.S:75]: This code doesn't touch any of the x86-64 callee-saved registers (which are `rsp`, `rbp`, `rbx`, and `r12`-`r15`).

**Pull request merged**